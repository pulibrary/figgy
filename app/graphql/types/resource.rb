# frozen_string_literal: true
module Types::Resource
  include Types::BaseInterface
  description "A resource in the system."
  orphan_types Types::ScannedResourceType, Types::FileSetType, Types::ScannedMapType

  field :id, String, null: true
  field :label, String, null: true
  field :viewing_hint, String, null: true
  field :url, String, null: true
  field :members, [Types::Resource], null: true
  field :source_metadata_identifier, String, null: true
  field :thumbnail, Types::Thumbnail, null: true

  definition_methods do
    def resolve_type(object, _context)
      "Types::#{object.class}Type".constantize
    end
  end

  def members
    @members ||= Wayfinder.for(object).members
  end

  def manifest_url
    helper.polymorphic_url([:manifest, object])
  end

  def url
    @url ||= helper.show_url(object)
  end

  # We need to centralize logic for navigating a MVW's members to find a
  # thumbnail file set. This is a hack to use the helper's logic for doing that.
  # refactor ticketed as https://github.com/pulibrary/figgy/issues/1708
  def helper
    @helper ||= ManifestBuilder::ManifestHelper.new.tap do |helper|
      helper.singleton_class.include(ThumbnailHelper)
      helper.define_singleton_method(:image_tag) do |url, _opts|
        url
      end
      helper.define_singleton_method(:image_path) do |url|
        url
      end
    end
  end

  def thumbnail
    return if object.try(:thumbnail_id).blank? || thumbnail_resource.blank?
    {
      id: thumbnail_resource.id.to_s,
      thumbnail_url: helper.figgy_thumbnail_path(thumbnail_resource),
      iiif_service_url: helper.figgy_thumbnail_path(thumbnail_resource).gsub("/full/!200,150/0/default.jpg", "")
    }
  end

  def thumbnail_resource
    @thumbnail_resource ||= query_service.find_by(id: object.try(:thumbnail_id).first)
  rescue Valkyrie::Persistence::ObjectNotFoundError
    nil
  end

  def query_service
    Valkyrie::MetadataAdapter.find(:indexing_persister).query_service
  end
end
