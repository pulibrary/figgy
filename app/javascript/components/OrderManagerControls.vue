<template>
  <wrapper class="lux-bg">
    <div class="lux-controls">
      <alert
        v-if="orderChanged"
        status="info"
      >
        Page order has changed.
      </alert>
      <input-button
        id="save_btn"
        variation="solid"
        size="medium"
        :disabled="isDisabled"
        @button-clicked="saveHandler($event)"
      >
        Apply Changes
      </input-button>
      <a
        v-if="!hidden"
        id="replace-file-button"
        :href="editLink"
      >
        Manage Page Files
      </a>
    </div>
    <heading
      v-if="selectedTotal === 1"
      level="h2"
    >
      Zoom <small>on the selected item</small>
    </heading>
    <div
      v-if="selectedTotal === 1"
      class="lux-osd-wrapper"
    >
      <div class="lux-osd">
        <div
          :id="viewerId"
          class="lux-viewport"
        />
      </div>
    </div>
  </wrapper>
</template>

<script>
import OpenSeadragon from 'openseadragon'
import { mapState } from 'vuex'
/**
 * This is the Persistence and Deep Zoom pieces of the Order Manager interface.
 * Note: use `yarn add openseadragon` for deep zoom to work.
 */
export default {
  name: 'Controls',
  status: 'ready',
  release: '1.0.0',
  type: 'Pattern',
  metaInfo: {
    title: 'OrderManager Controls',
    htmlAttrs: {
      lang: 'en'
    }
  },
  props: {
    /**
     * The html element name used for the component.
     */
    type: {
      type: String,
      default: 'div'
    },
    viewerId: {
      type: String,
      default: 'viewer'
    }
  },
  data: function () {
    return {
      viewer: null,
      osdId: this.viewerId
    }
  },
  computed: {
    ...mapState({
      resource: state => state.ordermanager.resource,
      gallery: state => state.gallery
    }),
    editLink: function () {
      let link = ''
      if (!this.hidden) {
        link = '/catalog/parent/' + this.resource.id + '/' + this.gallery.selected[0].id
      }
      return link
    },
    isMultiVolume: function () {
      return this.resource.isMultiVolume
    },
    isDisabled: function () {
      if (this.$store.getters.stateChanged) {
        return false
      } else {
        return true
      }
    },
    orderChanged: function () {
      return this.$store.getters.orderChanged
    },
    payloadFileset: function () {
      let changed = this.gallery.items.filter(item => this.gallery.changeList.indexOf(item.id) !== -1)
      let payload = changed.map(file => {
        return { id: file.id, title: file.title, page_type: file.viewingHint }
      })
      return payload
    },
    payloadVolume: function () {
      var changed = this.gallery.items.filter(item => this.gallery.changeList.indexOf(item.id) !== -1)
      var payload = changed.map(file => {
        return { id: file.id, title: file.title }
      })
      return payload
    },
    resourceClassName: function () {
      return this.resource.resourceClassName
    },
    selectedTotal () {
      return this.gallery.selected.length
    }
  },
  updated: function () {
    if (this.selectedTotal === 1) {
      this.initOSD()
    }
  },
  methods: {
    initOSD: function () {
      if (this.viewer) {
        this.viewer.destroy()
        this.viewer = null
      }
      this.viewer = OpenSeadragon({
        id: this.osdId,
        showNavigationControl: false,
        tileSources: [this.gallery.selected[0].service + '/info.json']
      })
    },
    hidden: function () {
      if (this.selectedTotal !== 1) {
        return true
      } else {
        return false
      }
    },
    galleryToFileset: function (items) {
      var members = items.filter(item => this.gallery.changeList.indexOf(item.id) > -1).map(item => {
        return { id: item.id, label: item.caption, viewingHint: item.viewingHint }
      })
      return members
    },
    galleryToResource: function (items) {
      var members = items.map(item => {
        return item.id
      })
      return members
    },
    saveHandler: function (event) {
      if (this.isMultiVolume) {
        this.saveMVW()
      } else {
        this.save()
      }
    },
    save: function () {
      let resource = {}
      resource.body = {
        id: this.resource.id,
        viewingDirection: this.resource.viewingDirection
          ? this.resource.viewingDirection.replace(/-/g, '').toUpperCase()
          : this.resource.viewingDirection,
        viewingHint: this.resource.viewingHint,
        startPage: this.resource.startCanvas,
        thumbnailId: this.resource.thumbnail,
        memberIds: this.galleryToResource(this.gallery.items)
      }
      resource.filesets = []
      let membersBody = this.galleryToFileset(this.gallery.items)
      let memberNum = membersBody.length
      for (let i = 0; i < memberNum; i++) {
        resource.filesets.push(membersBody[i])
      }
      window.resource = resource
      this.$store.dispatch('saveStateGql', resource)
    },
    saveMVW: function () {
      let body = {
        resource: {},
        volumes: this.payloadVolume
      }
      body.resource[this.resourceClassName] = {
        member_ids: this.imageIdList,
        viewing_direction: this.viewingDirection,
        thumbnail_id: this.thumbnail,
        id: this.id
      }
      this.$store.dispatch('saveStateGql', body)
    }
  }
}
</script>

<style lang="scss" scoped>
small {
  font-size: 1rem;
  font-weight: 400;
}
#replace-file-button {
  padding: 1.5rem;
  display: inline-block;
}
.lux-is-hidden {
  display: none;
}
.lux-bg {
  background: #f9f9f9;
  margin-left: -30px;
  margin-right: -30px;
  padding: 20px;
  display: flex;
  flex-direction: column;
  height: 100%;
}

.lux-osd {
  background: #fff;
  height: 100%;
  width: 100%;
}

.lux-osd-wrapper {
  background: #fff;
  flex-basis: 40%;
  border-radius: 4px;
  border: 2px solid #9ecaed;
  box-shadow: 0 0 10px #9ecaed;
  padding: 10px;
}

h3.lux-osd-title {
  font-size: 18px;
}

.lux-viewport {
  height: 100%;
  width: 100%;
}
</style>
