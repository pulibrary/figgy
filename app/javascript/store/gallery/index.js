export const galleryState = {
  items: [],
  selected: [],
  cut: [],
  changeList: [],
  ogItems: [],
}

export const galleryMutations = {
  CUT(state, itemArray) {
    state.cut = [...itemArray]
  },
  PASTE(state, itemArray) {
    state.items = [...itemArray]
  },
  SELECT(state, itemArray) {
    state.selected = [...itemArray]
  },
  SET_GALLERY(state, items) {
    state.items = items
    state.ogItems = items
  },
  SORT_ITEMS(state, value) {
    state.items = [...value]
  },
  UPDATE_CHANGES(state, changeList) {
    state.changeList = [...changeList]
  },
  UPDATE_ITEMS(state, items) {
    state.items = [...items]
  },
}

export const galleryModule = {
  state: galleryState,
  mutations: galleryMutations,
}
