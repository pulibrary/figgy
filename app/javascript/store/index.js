import Vue from 'vue/dist/vue.esm'
import Vuex from 'vuex'
import actions from './vuex/actions'
import {modules} from 'lux-design-system'
Vue.use(Vuex)

const store = new Vuex.Store({
  actions,
  modules: {
    ordermanager: modules.resourceModule,
    gallery: modules.galleryModule,
  }
})

export default store
