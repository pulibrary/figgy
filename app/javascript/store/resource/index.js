export const resourceState = {
  resource: {
    id: "",
    resourceClassName: "",
    bibId: "",
    label: "Resource not available.",
    thumbnail: "",
    startCanvas: "",
    isMultiVolume: false,
    viewingHint: null,
    viewingDirection: null,
    members: [],
    loadState: "NOT_LOADED",
    saveState: "NOT_SAVED",
    ogState: {},
  },
}

export const resourceMutations = {
  APPLY_STATE(state) {
    state.gallery.ogItems = state.gallery.items
    state.gallery.changeList = []
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
    state.resource.saveState = "NOT_SAVED"
  },
  CHANGE_RESOURCE_LOAD_STATE(state, loadState) {
    state.resource.loadState = loadState
  },
  SAVED_STATE(state, saveStatus) {
    state.resource.saveState = saveStatus
  },
  SET_RESOURCE(state, resource) {
    state.resource.id = resource.id
    state.resource.resourceClassName = resource.__typename
    state.resource.label = resource.label
    state.resource.startCanvas = resource.startPage
    state.resource.viewingHint = resource.viewingHint
    state.resource.viewingDirection = resource.viewingDirection
    state.resource.thumbnail = resource.thumbnail != null ? resource.thumbnail.id : null
    state.resource.members = resource.members

    const items = resource.members.map(member => ({
      id: member.id,
      viewingHint: member.viewingHint != null ? member.viewingHint : "single",
      caption: member.label, // member.__typename + " : " + member.id,
      service:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl
          : Global.figgy.resource.defaultThumbnail,
      mediaUrl:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl + "/full/300,/0/default.jpg"
          : Global.figgy.resource.defaultThumbnail,
    }))
    state.gallery.items = items
    state.gallery.ogItems = items
    state.resource.loadState = "LOADED"
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail != null ? resource.thumbnail.id : null,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
  },
  UPDATE_STARTCANVAS(state, startCanvas) {
    state.resource.startCanvas = startCanvas
  },
  UPDATE_THUMBNAIL(state, thumbnail) {
    state.resource.thumbnail = thumbnail
  },
  UPDATE_VIEWDIR(state, viewDir) {
    state.resource.viewingDirection = viewDir
  },
  UPDATE_VIEWHINT(state, viewHint) {
    state.resource.viewingHint = viewHint
  },
}

export const resourceGetters = {
  getMemberCount: state => {
    return state.resource.members.length
  },
  isMultiVolume: state => {
    const volumes = state.resource.members.filter(member => member.__typename === "ScannedResource")
    return volumes.length > 0 ? true : false
  },
  orderChanged: state => {
    let ogOrder = JSON.stringify(state.gallery.ogItems.map(item => item.id))
    let imgOrder = JSON.stringify(state.gallery.items.map(item => item.id))
    return ogOrder !== imgOrder
  },
  stateChanged: (state, getters) => {
    var propsChanged = []
    propsChanged.push(state.resource.ogState.thumbnail !== state.resource.thumbnail)
    propsChanged.push(state.resource.ogState.startCanvas !== state.resource.startCanvas)
    propsChanged.push(state.resource.ogState.viewingHint !== state.resource.viewingHint)
    propsChanged.push(state.resource.ogState.viewingDirection !== state.resource.viewingDirection)
    propsChanged.push(state.gallery.changeList.length > 0)
    propsChanged.push(getters.orderChanged)
    if (propsChanged.indexOf(true) > -1) {
      return true
    } else {
      return false
    }
  },
}
