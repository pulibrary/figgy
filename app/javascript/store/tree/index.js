export const treeState = {
  selected: null,
}

export const treeMutations = {
  SELECT(state, selectedId) {
    state.selected = selectedId
  },
}
