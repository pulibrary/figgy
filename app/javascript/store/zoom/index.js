export const zoomState = {
  zoomed: null,
}

export const zoomMutations = {
  ZOOM(state, obj) {
    state.zoomed = obj
  },
  RESET_ZOOM(state) {
    state.zoomed = null
  },
}

export const zoomGetters = {
  getIsZoomed: state => {
    return state.zoomed
  },
}
