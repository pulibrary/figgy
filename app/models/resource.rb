# frozen_string_literal: true
class Resource < Valkyrie::Resource
  def self.human_readable_type
    default = @_human_readable_type || name.demodulize.titleize
    I18n.translate("models.#{new.model_name.i18n_key}", default: default)
  end

  def self.can_have_manifests?
    false
  end

  def human_readable_type
    self.class.human_readable_type
  end

  def self.model_name
    ::ActiveModel::Name.new(self)
  end

  # Determines if this is an image resource
  # @return [TrueClass, FalseClass]
  def image_resource?
    false
  end

  # Determines if this is a geospatial resource
  # @return [TrueClass, FalseClass]
  def geo_resource?
    false
  end

  # Determines if this is a media resource
  # @return [TrueClass, FalseClass]
  def media_resource?
    false
  end

  # Virtual property used for stashing pre-loaded objects. Populated by a query.
  attr_accessor :loaded
  def loaded
    @loaded ||= {}
  end
end
