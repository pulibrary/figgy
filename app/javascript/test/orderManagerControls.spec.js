import Vuex from "vuex"
import { createLocalVue, mount, shallowMount } from "@vue/test-utils"
import OrderManagerControls from "../components/OrderManagerControls.vue"
import { resourceMutations, resourceGetters } from "../store/resource"
import { modules } from 'lux-design-system'

// create an extended `Vue` constructor
const localVue = createLocalVue()
localVue.use(Vuex)

let wrapper
let getters
let options
let store
let actions
let items = [
  { id: "a", title: "A", caption: "a", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "b", title: "B", caption: "b", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "c", title: "C", caption: "c", mediaUrl: "https://picsum.photos/600/300/?random" },
]
let resourceObject = {
  id: "aea40813-e0ed-4307-aae9-aec53b26bdda",
  label: "Resource with 3 files",
  viewingHint: "individuals",
  viewingDirection: "LEFTTORIGHT",
  startPage: "8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
  thumbnail: {
    id: "8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
    thumbnailUrl:
      "http://localhost:3000/image-service/8ffd7a03-ec0e-46c1-a347-e4b19cb7839f/full/!200,150/0/default.jpg",
    iiifServiceUrl: "http://localhost:3000/image-service/8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
  },
  __typename: "ScannedResource",
  members: [
    {
      id: "8ffd7a03-ec0e-46c1-a347-e4b19cb7839f",
      label: "a",
      viewingHint: null,
      thumbnail: {
        iiifServiceUrl:
          "https://libimages1.princeton.edu/loris/figgy_prod/f7%2F67%2Ffe%2Ff767fe4247524c5f96e16eba2ff93301%2Fintermediate_file.jp2",
      },
      __typename: "FileSet",
    },
    {
      id: "8f0a0908-317f-414e-a78a-c38a4a3b28e3",
      label: "b",
      viewingHint: null,
      thumbnail: {},
      __typename: "FileSet",
    },
    {
      id: "ea01019e-f644-4416-b99c-1b44bf49d060",
      label: "c",
      viewingHint: null,
      thumbnail: {
        iiifServiceUrl:
          "https://libimages1.princeton.edu/loris/figgy_prod/d9%2Fb5%2F8c%2Fd9b58c8f3e554706bec4d977b12cd4e4%2Fintermediate_file.jp2",
      },
      __typename: "FileSet",
    },
  ],
}

describe("OrderManagerControls.vue", () => {
  beforeEach(() => {
    actions = {
      loadImageCollectionGql: vi.fn(),
      saveStateGql: vi.fn()
    }

    const gallery = {
      state: {
        items: items,
        selected: [items[0]],
        cut: [],
        changeList: [],
        ogItems: items,
      },
      mutations: modules.galleryModule.mutations,
    }

    const getters = {
      orderChanged: () => false,
      stateChanged: () => false
    }

    const resource = {
      state: {
        resource: {
          id: "",
          resourceClassName: "",
          bibId: "",
          label: "Resource not available.",
          thumbnail: "",
          startCanvas: "",
          isMultiVolume: false,
          viewingHint: null,
          viewingDirection: null,
          members: [],
          loadState: "NOT_LOADED",
          saveState: "NOT_SAVED",
          ogState: {},
        },
      },
      mutations: resourceMutations,
      getters: getters,
      actions: actions,
      modules: {
        gallery: gallery,
      },
    }

    store = new Vuex.Store({
      modules: {
        ordermanager: resource,
        gallery: gallery,
      },
    })

    options = {
      computed: {
        orderChanged: function () {
          return getters.orderChanged
        },
        isDisabled: function () {
          if (getters.stateChanged) {
            return false
          } else {
            return true
          }
        }
      }
    }

    wrapper = mount(OrderManagerControls, {
      options,
      localVue,
      store,
      stubs: ["heading", "input-button", "wrapper"],
    })
  })

  it('will not allow save when nothing has changed', () => {
    expect(wrapper.vm.isDisabled).toBeTruthy()
  })

  it('shows the openseadragon div when a single fileset is selected', () => {
    expect(wrapper.find('heading-stub').exists()).toBe(true)
    expect(wrapper.find('.lux-osd-wrapper').exists()).toBe(true)
  })

  it('displays the Manage Page Files button when one item is selected', () => {
    expect(wrapper.vm.hidden()).toBe(false)
  })

  it('isMultiVolume is false when it is not multi-volume', () => {
    expect(wrapper.vm.isMultiVolume).toBe(false)
  })

  it('maps gallery items to resource, to prepare data for save', () => {
    expect(wrapper.vm.galleryToResource(items)).toEqual(["a", "b", "c"])
  })

  it('tests a number of scenarios once something has changed', async () => {
    actions = {
      loadImageCollectionGql: vi.fn(),
      saveStateGql: vi.fn()
    }

    const changedGetters = {
      orderChanged: () => true,
      stateChanged: () => true
    }

    let changedItems = [
      { id: "c", title: "C", caption: "c", mediaUrl: "https://picsum.photos/600/300/?random" },
      { id: "a", title: "A", caption: "a", mediaUrl: "https://picsum.photos/600/300/?random" },
      { id: "b", title: "B", caption: "b", mediaUrl: "https://picsum.photos/600/300/?random" },
    ]

    const changedGallery = {
      state: {
        items: changedItems,
        selected: [items[0]],
        cut: [],
        changeList: [],
        // changeList: ["2"],
        ogItems: items,
      },
      mutations: modules.galleryModule.mutations,
    }

    const changedResource = {
      state: {
        resource: {
          id: "",
          resourceClassName: "",
          bibId: "",
          label: "Resource not available.",
          thumbnail: "",
          startCanvas: "",
          isMultiVolume: false,
          viewingHint: null,
          viewingDirection: null,
          members: [],
          loadState: "NOT_LOADED",
          saveState: "SAVING",
          ogState: {},
        },
      },
      mutations: resourceMutations,
      getters: changedGetters,
      actions: actions,
      modules: {
        gallery: changedGallery,
      },
    }

    store = new Vuex.Store({
      modules: {
        ordermanager: changedResource,
        gallery: changedGallery,
      },
    })

    options = {
      computed: {
        orderChanged: function () {
          return getters.orderChanged
        },
        isDisabled: function () {
          if (getters.stateChanged) {
            return false
          } else {
            return true
          }
        }
      }
    }

    wrapper = mount(OrderManagerControls, {
      options,
      localVue,
      store,
      stubs: ["alert", "heading", "input-button", "wrapper"],
    })

    // orderChanged should be true when items and ogItems don't match
    expect(wrapper.vm.orderChanged).toBe(true)

    // displays an alert when page order has changed
    expect(wrapper.find('alert-stub').exists()).toBe(true)

    // Disable the button while saving.
    expect(wrapper.vm.isDisabled).toBeTruthy()

    // calls the appropriate action on save
    await wrapper.vm.saveHandler()
    expect(actions.saveStateGql).toHaveBeenCalled()
  })

  it('assures saveStateGql is called for MVWs', () => {
    actions = {
      loadImageCollectionGql: vi.fn(),
      saveStateGql: vi.fn()
    }

    const changedGetters = {
      orderChanged: () => true,
      stateChanged: () => true
    }

    let changedItems = [
      { id: "c", title: "C", caption: "c", mediaUrl: "https://picsum.photos/600/300/?random" },
      { id: "a", title: "A", caption: "a", mediaUrl: "https://picsum.photos/600/300/?random" },
      { id: "b", title: "B", caption: "b", mediaUrl: "https://picsum.photos/600/300/?random" },
    ]

    const changedGallery = {
      state: {
        items: changedItems,
        selected: [items[0]],
        cut: [],
        changeList: [],
        // changeList: ["2"],
        ogItems: items,
      },
      mutations: modules.galleryModule.mutations,
    }

    const changedResource = {
      state: {
        resource: {
          id: "",
          resourceClassName: "",
          bibId: "",
          label: "Resource not available.",
          thumbnail: "",
          startCanvas: "",
          isMultiVolume: true,
          viewingHint: null,
          viewingDirection: null,
          members: [],
          loadState: "NOT_LOADED",
          saveState: "NOT_SAVED",
          ogState: {},
        },
      },
      mutations: resourceMutations,
      getters: changedGetters,
      actions: actions,
      modules: {
        gallery: changedGallery,
      },
    }

    store = new Vuex.Store({
      modules: {
        ordermanager: changedResource,
        gallery: changedGallery,
      },
    })

    options = {
      computed: {
        orderChanged: function () {
          return getters.orderChanged
        },
        isDisabled: function () {
          if (getters.stateChanged) {
            return false
          } else {
            return true
          }
        }
      }
    }

    wrapper = mount(OrderManagerControls, {
      options,
      localVue,
      store,
      stubs: ["alert", "heading", "input-button", "wrapper"],
    })

    // calls the appropriate action on save
    wrapper.vm.saveHandler()
    expect(actions.saveStateGql).toHaveBeenCalled()
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
