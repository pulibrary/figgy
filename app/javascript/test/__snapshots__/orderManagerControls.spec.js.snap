// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderManagerControls.vue has the expected html structure 1`] = `
<wrapper-stub
  class="lux-bg"
>
  <div
    class="lux-controls"
  >
    <!---->
     
    <input-button-stub
      disabled="true"
      id="save_btn"
      size="medium"
      variation="solid"
    >
      
      Apply Changes
    
    </input-button-stub>
     
    <!---->
  </div>
   
  <heading-stub
    level="h2"
  >
    
    Zoom 
    <small>
      on the selected item
    </small>
  </heading-stub>
   
  <div
    class="lux-osd-wrapper"
  >
    <div
      class="lux-osd"
    >
      <div
        class="lux-viewport"
        id="viewer"
      />
    </div>
  </div>
</wrapper-stub>
`;
